### ISSUE TITLE
Why Logger::interpolate() is private ???

### ISSUE THREAD

**Issue by thochra** [bzikarsky/gelf-php#92]
Hello,
I'm switching from an old GELF lib to yours and I needed to do some adjustments regarding our needs.

I was wondering if is there is any interest of having this following method to be private:
```php
    private static function interpolate($message, array $context)
    {
        // build a replacement array with braces around the context keys
        $replace = array();
        foreach ($context as $key => $val) {
            $replace['{' . $key . '}'] = $val;
        }

        // interpolate replacement values into the message and return
        return strtr($message, $replace);
    }
```

Instead of simply overriding the interpolate like that:
```php
    protected static function interpolate($message, array $context)
    {
        // Do some stufff
       return parent::interpolate($message, $context);
    }
```

I have to override the `protected function initMessage($level, $message, array $context)` method :/

**Comment by bzikarsky** [comment]
In general: I'm quite a fan of "Keep as much stuff private as possible - Opening up is always possible, closing down break the API". So that's why this private. 

Specific to your usecase: The interpolation method is a straight copy from the PSR standard. Its presence is primarily for convenience. The concept of interpolating a bunch of context parameters into a logging-string is not really a concept I want to expose on the library's API. Does it hurt you much?

**Comment by Gummibeer** [comment]
I think the goal isn't `public` but `protected`. I'm a fan of `protected` in libraries to allow others to customize the inner behavior by overriding the class.

**Comment by thochra** [comment]
@Gummibeer Yes it's exactly the point!

It gives us (frameworks users) more flexibility to adapt a functionality for a specific context without having to do extra overrides that would not be necessary + avoid crapy copy paste of the the private code that we'll have to update manually with the framework updates.

I almost never meet the requirement of keeping something `private` compared to `protected`.

That said, I dealt with it with not so much effort as the `interpolate` is not widely used and I did not had to override other stuff.

**Comment by bzikarsky** [comment]
It really doesn't matter whether its public or protected. If a library method can be used by the library's users its part of the libraries API contract: You can't change its signature, its calling-context, or anything else about the method without bumping the library's major version.

**Comment by Gummibeer** [comment]
@bzikarsky but this just counts for changes in the original repo. If I adjust something in my own implementation I add another layer on top of the library. So for sure I can make the library, with a feature release, unusable in my app cause my changes conflict with library ones. But that's my problem and I can fix it also without touching the library itself.
If you make a method `private` it's impossible to adjust this behavior in any way.

That's why most libraries just use `public` *documented API & methods that can cause breaking changes with a new major version* and `protected` *internal library logic that will not cause a new major version even if they are removed or changed* but could be overridden by library users in there own implementation if they need (without warranty and by their own risk).

So yeah, I can do something like (following) and break the whole library - but it's just my implementation and I will know where to go first if something is broken.
```php
protected function interpolate()
{
    throw new Exception('I do not like this method.);
}
```

**Comment by bzikarsky** [comment]
Maybe I have to make this more clear.

When I provide a library, I provide an API contract. This contract is versioned with [semantic versioning](https://semver.org/). Part of this contract is (especially in the PHP world) every signature you can use.

The most obvious ones are the class-constructor and method signatures. But also part of the contract is the inheritance API. If I don't flag a class as `final` I allow every library-consumer to extend the class, therefore every `protected` method is then also part of API under contract.

SemVer states very clearly that there can't be any signature breaking change without a major version bump. 

In conclusion: if I change (the signature, the existence, the semantic) of a protected method, in a non-final class, I need to bump the major version.

A possible workaround is an additional flag (like an @Internal annotation) which requires documentation.

**Comment by Gummibeer** [comment]
@bzikarsky I that exactly this is the point of discussion - semver just talks about the `public` API. For me and a lot of other library developers this is that the collection of methods and properties that are flagged as `public` - cause only these could be called without adding a new class or changing the code in any other way.
`protected` methods need a new class on top to change them, after this the code I call isn't the library code anymore.
It's the same for `public` ones, I also can put a class on top and change everything the public methods do - but after this it's my own implementation of your library and no more your documented public API.

The the methods that could be changed by other developers aren't the same like the ones that are part of your semver public API contract.

Part of this is everything I can call by just doing `require your/library` and `(new LibraryClass())->...` everything else I can do with your library isn't part of this contract. That's why changes on protected and private methods in other libraries aren't BC cause they just contain internal logic that could be overridden but if they are it's not part of the library owners API contract and cause of this they could be changed by simple minor releases.

**Comment by bzikarsky** [comment]
I think I made my position clear. We disagree on which parts of an API are part of the contract. I will therefore close this issue.

Thanks for taking time to discuss this.
